File: py_drill/Problems/ARH05_ValidAnagram.py
# REF : https://leetcode.com/problems/valid-anagram/
# NOTES :
#       Hash map of character & their count
#       dict.get() is exception safe
#       sorting can also be done but O(nlogn)

# T: O(n)
# S: O(n)

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        hash_s, hash_t = {}, {}
        for i in range(len(s)):
            hash_s[s[i]] = hash_s.get(s[i], 0) + 1
            hash_t[t[i]] = hash_t.get(t[i], 0) + 1

        for c in hash_s:
            if hash_s[c] != hash_t.get(c, 0):
                return False

        return True

class Solution2:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        hashmap = {}
        for i in range(len(s)):
            hashmap[s[i]] = hashmap.get(s[i], 0) + 1    # increment +1 for char in hashmap
            hashmap[t[i]] = hashmap.get(t[i], 0) - 1    # decrement -1 for char in hashmap

        for ele in hashmap.values():    # all values(counts or occurances of char) of hashmap will be zero
            if ele:                     # if both strings are anagram of each other otherwise some non-zero
                return False            # will be there
        return True

def main():
    print("output : {}".format(Solution().isAnagram("anagram", "nagaram")))
    print("output : {}".format(Solution2().isAnagram("anagram", "nagaram")))

if __name__ == "__main__":
    main()

File: cpp_drill/Problems/ARH05_ValidAnagram.cpp
// REF : https://leetcode.com/problems/valid-anagram/

#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size())
            return false;

        unordered_map<char, int> countS;
        unordered_map<char, int> countT;

        for (int i = 0; i < s.size(); i++) {
            countS[s[i]]++;
            countT[t[i]]++;
        }

        for (auto &[key, value] : countS) {
            try {
                if (value != countT.at(key))    // try catch this countT could raise exception
                    return false;
            }
            catch(const exception& e) {
                cerr << e.what() << '\n';
                return false;
            }
        }

        return true;
    }
};

class Solution2 {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size())
            return false;

        unordered_map<char, int> hashmap;

        for (int i = 0; i < s.size(); i++) {
            hashmap[s[i]]++;
            hashmap[t[i]]--;
        }

        for (auto &[key, value] : hashmap) {
            if (value != 0)
                return false;
        }

        return true;
    }
};

int main() {
    cout << "hello" << endl;
    cout << "output : " << Solution().isAnagram("anagram", "nagaram") << endl;
    cout << "output : " << Solution2().isAnagram("anagram", "nagaram") << endl;
    return 0;
}